### PR TITLE
Added script for encrypting and uploading data to s3

### PR DESCRIPTION
**Pull request by arkal**
1. Script can upload files or directories to s3 with or without encryption. 
2. Supports specification of pseudo directories in the bucket to store the  data.
3. Retains local directory structure in the bucket.

Shoutout to @jvivian who wrote the original script.

### CODE REVIEW

File: contrib/encrypt_files_in_dir_to_s3.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    S3.  It requires a master key to encrypt the datum with, and a bucket to
+    drop the results into.  If remote dir is set, the datum is dropped into the
+    provided directory.
+    datum - :str: Path to file or folder to write to s3am
+    master_key - :str: Path to master key
+    bucket - :str: Bucket of s3am
+    remote_dir - :str: Path describing pseudo dir to store files on s3
+    '''
+    exit_codes = []
+    s3_url_base = 'https://s3-us-west-2.amazonaws.com/'
```
Comment (hannes-ucsc) [inline contrib/encrypt_files_in_dir_to_s3.py]: The region part of the URL should be configurable.

Comment (arkal) [inline contrib/encrypt_files_in_dir_to_s3.py]: It looks like this script was working with an old version of s3am that was runs as

```
s3am upload [options] SRC_URL BUCKET DST_DIR
```

s3am now runs as 

```
s3am upload [options] SRC_URL DST_DIR
```

where `DST_URL` is of the form `S3://BUCKET/DST_FILENAME`.

Region will be used at the moment only in the creation of the file specific key. 

When are we porting to the new encryption method -- where we will use a url similar to above mentioned DST_URL for encryption?

Comment (hannes-ucsc) [inline contrib/encrypt_files_in_dir_to_s3.py]: Sorry, I don't understand the question. Let's talk about this during scrum.

File: contrib/encrypt_files_in_dir_to_s3.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    :param str bucket_name: Name of an S3 bucket
+    :param dict aws_creds: Credentials to use for the query
+    :return: Region hosting the bucket
+    """
+    conn = S3Connection(aws_creds['aws_access_key_id'], aws_creds['aws_secret_access_key'])
+    bucket = conn.lookup(bucket_name)
+    if not bucket:
+        # TODO: Make the bucket here?
+        raise InputParameterError('The provided bucket %s doesn\'t exist' % bucket_name)
+    return str(bucket.get_location())
```
Comment (hannes-ucsc) [inline contrib/encrypt_files_in_dir_to_s3.py]: us-east-1 comes back as empty string.

https://github.com/BD2KGenomics/s3am/blob/master/src/s3am/operations.py#L110

File: contrib/encrypt_files_in_dir_to_s3.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+
+def get_bucket_region(bucket_name, aws_creds):
+    """
+    Queries S3 to identify the region hosting the provided bucket.
+    :param str bucket_name: Name of an S3 bucket
+    :param dict aws_creds: Credentials to use for the query
+    :return: Region hosting the bucket
+    """
+    conn = S3Connection(aws_creds['aws_access_key_id'], aws_creds['aws_secret_access_key'])
```
Comment (hannes-ucsc) [inline contrib/encrypt_files_in_dir_to_s3.py]: Remove explicit credentials. Rely on boto's default mechanisms for configuring credentials.

Comment (arkal) [inline contrib/encrypt_files_in_dir_to_s3.py]: Oops didn't know boto did that.

File: contrib/encrypt_files_in_dir_to_s3.py (hunk 2 above)
Comment (arkal) [inline contrib/encrypt_files_in_dir_to_s3.py]: Will fix.

Comment (arkal) [inline contrib/encrypt_files_in_dir_to_s3.py]: So looking at [this](http://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region),  the http endpoints for us-east-1 are `s3.amazonaws.com` and `s3-external-1.amazonaws.com` so i'm going to return `external-1`.

**Comment by arkal** [comment]
If the user specifies a directory, there will very possibly be tens to hundreds of files uploaded to the same bucket all sharing the same s3 url base (`S3://bucketname/`) and https url base (`https://s3<ENDPOINT>.amazonaws.com/`).

I created a BucketInfo class that queries amazon for the bucket location, then sets up the s3 and https url bases. the bucket has 2 methods to get the full https- and s3 urls for a given key.

**Comment by hannes-ucsc** [comment]
As discussed offline, we will take a different approach to these. See #30, #31, #32, #33, #34.

**Pull request closed without merging**